Transaction details page

song50119 opened the pull request:
### Resolves

Issue #63

### What's new

- Create transaction details page
- Click on overview/transactions page's transaction to view transaction details
### Relevant screenshots or logs

<img width="912" alt="Screen Shot 2020-05-01 at 22 59 52" src="https://user-images.githubusercontent.com/9800579/80819628-81deb900-8bff-11ea-8925-6b2b8b3dd16a.png">

itswisdomagain commented:
Is there another qr code library that does what we want - print smaller images of qr code to terminal?

Jujhar commented:
Changed to terminalprompt and found as marginlly smaller QR output as this one has no outside padding.

Not sure how to remove those blank lines any ideas?
 A part of me feels like those are fine and Radeah was talking about double extra lines rather than just one.

itswisdomagain commented on cli/handlers.go:
```diff
@@ -94,6 +97,19 @@ func send(walletrpcclient *rpcclient.Client, custom bool) (*response, error) {
+  text, _ := reader.ReadString('\n')
+
+	if (strings.TrimRight(text, "\n") == "n") {
+		return &response{}, nil
```
> Changed to terminalprompt and found a marginlly smaller QR output as this one has no outside padding.

Not small enough. You had another idea of using a graphics library, discuss that with @raedah 

> Not sure how to remove those blank lines any ideas?

Look into the `terminalprompt` package

> A part of me feels like those are fine and Radeah was talking about double extra lines rather than just one.

Nah. The blank lines have no use

itswisdomagain replied:
That's because `RunCommand` in `cli.go` in line 83 uses the return value as is. You should return nil and in main.go, check if res is nil before attempting to print. Your main.go should look like this:

```
if res != nil {
     printResult(tabWriter(os.Stdout), res)
}
```

Jujhar replied:
No longer an issue

itswisdomagain commented:
> Changed to terminalprompt and found a marginlly smaller QR output as this one has no outside padding.

Not small enough. You had another idea of using a graphics library, discuss that with @raedah 

> Not sure how to remove those blank lines any ideas?

Look into the `terminalprompt` package

> A part of me feels like those are fine and Radeah was talking about double extra lines rather than just one.

Nah. The blank lines have no use

itswisdomagain commented:
I'm putting this PR as WIP. After discussing with @raedah on how to deal with the mighty qrcode image on terminal, and fixing other issues highlighted, remove the [WIP] label

Jujhar commented:
Fixed and ready to go.

![qr-code-72](https://user-images.githubusercontent.com/5521110/50543912-d8665e00-0b99-11e9-8fcb-373c256a0b6f.png)

Jujhar commented:
Updating to add optional QR display via additional prompt.

itswisdomagain commented:
@Jujhar rebase your PR branch with master, fix conflicts and then push. Ensure you give attention to ALL previous review comments (there's a comment about using `nil` instead of `&response{}` that you haven't attended to)

When you've done that, rename this PR to remove the [WIP] flag, we don't want to spend time reviewing the code yet until those are done.

itswisdomagain commented on cli/commands/helpers.go:
```diff
@@ -63,7 +63,11 @@ func getSendDestinationAddress(c *walletrpcclient.Client) (string, error) {
 
 		if !isValid {
-			return errors.New("invalid address")
+			if (address == "") {
```
This check should be done before line 60. There's no need to attempt to validate an empty address

itswisdomagain commented on cli/commands/helpers.go:
```diff
@@ -85,6 +89,10 @@ func getSendAmount() (float64, error) {
 	validateAmount := func(input string) error {
 		amount, err = strconv.ParseFloat(input, 64)
 		if err != nil {
+			if (input == "") {
```
Always check for empty input before trying to "process" the input. This check should be done before line 90

itswisdomagain commented on cli/commands/receive.go:
```diff
@@ -42,21 +43,22 @@ func (r ReceiveCommand) Run(client *walletrpcclient.Client, args []string) error
+	validateConfirm := func(address string) error {
+		return  nil
 	}
+	confirm, _ := terminalprompt.RequestInput("Would you like to a generate QR code? (y/n) ", validateConfirm)
```
If `validateConfirm` doesn't do anything, you can simply pass nil here instead.

itswisdomagain commented on cli/commands/receive.go:
```diff
@@ -42,21 +43,22 @@ func (r ReceiveCommand) Run(client *walletrpcclient.Client, args []string) error
-	if err != nil {
-		return fmt.Errorf("Error generating QR Code: %s", err.Error())
+	// Print out address as string
+	fmt.Printf("%s\n", receiveResult.Address)
```
`fmt.Println(receiveResult.Address)`

itswisdomagain commented on cli/commands/receive.go:
```diff
@@ -42,21 +43,22 @@ func (r ReceiveCommand) Run(client *walletrpcclient.Client, args []string) error
-	columns := []string{
-		"Address",
-		"QR Code",
+	if confirm == "Yes" || confirm == "yes" {
```
Do case-insensitive comparison using `strings.EqualFold(confirm, "yes")`
Also, no need to assign `confirm = "y"` only to check the value of confirm again in the next line. Unnecesary.
Do all your comparison here:
`if strings.EqualFold(confirm, "yes") || strings.EqualFold(confirm, "y")`, then display the qrcode

itswisdomagain commented on cli/commands/send.go:
```diff
@@ -74,6 +74,22 @@ func send(rpcclient *walletrpcclient.Client, custom bool) error {
+	if confirm != "y" {
+		fmt.Printf("Operation cancelled\n")
+		return  nil
+	}
```
The comments above apply to this file as well

Jujhar commented:
Done

itswisdomagain commented on cli/commands/helpers.go:
```diff
@@ -83,6 +87,10 @@ func getSendAmount() (float64, error) {
 
 	validateAmount := func(input string) error {
+		if (input == "") {
+			return errors.New("That is did a valid amount. Try again.")
```
Typo here

itswisdomagain replied:
Also, the message here should not be "invalid amount", but "Amount not specified" like was done in the address validate function

itswisdomagain commented on cli/commands/helpers.go:
```diff
@@ -83,6 +87,10 @@ func getSendAmount() (float64, error) {
+
 		amount, err = strconv.ParseFloat(input, 64)
 		if err != nil {
 			return fmt.Errorf("error parsing amount: %s", err.Error())
```
This should be "Invalid amount. Try again"

itswisdomagain replied:
This isn't resolved

itswisdomagain commented on cli/commands/send.go:
```diff
@@ -86,17 +96,6 @@ func send(rpcclient *walletrpcclient.Client, custom bool) error {
-	}
-
-	termio.PrintTabularResult(termio.StdoutWriter, columns, rows)
+	fmt.Println("sent. txid %s", result.TransactionHash)
```
This should be `fmt.Printf("Sent. Txid %s", result.TransactionHash)` or `fmt.Println("Send. Txid", result.TransactionHash)`

Jujhar commented:
Alright updated again.

itswisdomagain commented on cli/commands/send.go:
```diff
@@ -72,7 +76,34 @@ func send(wallet walletcore.Wallet, custom bool) (err error) {
+		confirm = "y"
+	}
+
+	if confirm != "y" {
```
Instead of checking `if confirm != "y"` here, you can check `if strings.EqualFold(confirm, "n")` in line 96 and cancel the operation there instead.

Also note that the `validateConfirm` function will not accept "yes" or "no", only "y" or "n"

itswisdomagain commented on cli/commands/send.go:
```diff
@@ -72,7 +76,34 @@ func send(wallet walletcore.Wallet, custom bool) (err error) {
+	}
+
+	if confirm != "y" {
+		fmt.Printf("Operation cancelled\n")
```
`fmt.Println("Operation cancelled")`

Jujhar commented:
Fixed

itswisdomagain commented on cli/commands/send.go:
```diff
@@ -72,7 +76,30 @@ func send(wallet walletcore.Wallet, custom bool) (err error) {
+			return fmt.Errorf("invalid option, try again")
+		}
+	}
+	confirm, err := terminalprompt.RequestInput("Are you sure? (y/N) ", validateConfirm)
```
As seen here https://github.com/raedahgroup/godcr/issues/54#issuecomment-446785257, there is no default option for this prompt. User must enter "y" or "n" to proceed. This prompt should not have a capital option (capital options imply that if the user does not enter any value, it's assumed they accept the option in capital)

This should be "Are you sure? (y/n)". The space at the end of that message isn't needed as well

itswisdomagain commented on cli/commands/send.go:
```diff
@@ -72,7 +76,30 @@ func send(wallet walletcore.Wallet, custom bool) (err error) {
+	validateConfirm := func(userResponse string) error {
+		userResponse = strings.TrimSpace(userResponse)
+		userResponse = strings.Trim(userResponse, `"`)
+		if userResponse == "" || strings.EqualFold("Y", userResponse) || strings.EqualFold("n", userResponse) {
```
This validate function should not accept empty string values ("") since there is no default option. User must enter "y" or "n". See comment below for more

itswisdomagain commented on cli/commands/receive.go:
```diff
@@ -44,21 +47,30 @@ func (receiveCommand ReceiveCommand) Run(ctx context.Context, wallet walletcore.
-	columns := []string{
-		"Address",
-		"QR Code",
+	confirm, err := terminalprompt.RequestInput("Would you like to generate a QR code? (y/N) ", validateConfirm)
```
The trailing space at the end of this prompt message is not needed

itswisdomagain commented on cli/commands/send.go:
```diff
@@ -72,7 +76,30 @@ func send(wallet walletcore.Wallet, custom bool) (err error) {
 		return err
 	}
 
+	fmt.Println("You are about to send %f DCR to %s", sendAmount, destinationAddress)
```
`fmt.Printf("You are about to send %f DCR to %s\n", sendAmount, destinationAddress)`

Commit:
wallet: change struct RecentTransaction to Transaction
- Transactions.Txs slice of dcrlibwallet.Transaction to slice of Transaction

Commit:
ui: create transaction details page
- Add transaction widgets for layouts and inputs/outputs
- Handler transaction inputs widgets
- Implement click able to transaction on "overview page"
- window add `walletTransaction  *wallet.Transaction` contains transaction info

Commit:
ui: Update transaction details page
- Show the page as modal
- Change transaciton_page.go to transaction.go
- Add cancel button at header
- Change "hideTransactionDetails" widgets to "viewTxnOnDcrdata"

Commit:
ui: update transaction details

- Display transaction confirmations number

- Click on "View on dcrdata" button will open txn on browser

- Add click event to transaction item on transaction_page.go

Commit:
ui: update handler input toTransactionDetails click

Commit:
Rebase to master

oshorefueled commented on ui/handle_input.go:
```diff
@@ -418,3 +452,31 @@ func (win *Window) sortTransactions() {
 	})
 }
+
+func (win *Window) updateToTransactionDetailsBtns() {
```
you can make the "Btn" in the method name fully spelt as in "updateToTransactionDetailsButtons"

oshorefueled commented on ui/transaction.go:
```diff
@@ -0,0 +1,225 @@
+	transactionOutputsContainer = &layout.List{Axis: layout.Vertical}
+)
+
+func (win *Window) TransactionPage() {
```
This isn't a page but a dialog, appropriate names should be used to avoid confusion. So this would be TransactionDetailDialog. The file name should be renamed as well, "transaction_detail_dialog" would work better than just "transaction".

song50119 replied:
Okay

oshorefueled commented on ui/handle_input.go:
```diff
@@ -418,3 +452,31 @@ func (win *Window) sortTransactions() {
 }
+
+func (win *Window) updateToTransactionDetailsBtns() {
+	win.inputs.toTransactionDetails = make(map[string]*gesture.Click)
```
since you know the size of the transactions, you can declare the size of the map.

oshorefueled commented on ui/handle_input.go:
```diff
@@ -418,3 +452,31 @@ func (win *Window) sortTransactions() {
+
+func viewTxnOnBrowser(hash string) {
+	var err error
+	url := "https://testnet.dcrdata.org/tx/" + hash
```
you can add a "getBlockExplorerUrl" method to the wallet package. The method should return the url with respect to the network the app is running on.

Commit:
wallet: add Total to Transactions struct

- Add GetBlockExplorerURL

Commit:
ui: change name TransactionPage > TransactionDetailDialog
- Change updateToTransactionDetailsBtns > updateToTransactionDetails
- Init toTransactionDetails widget if Transactions.Total > 0
- Fix lint errors

Commit:
ui: change updateToTransactionDetails -> updateToTransactionDetailsButtons

Commit:
ui: GetAllTransactions when new block attached

Commit:
wallet: Update transaction listener
- Add "GetTransaction" to get transaction details
- Add const "BlockAttached"
- Update "txlistener.go", OnBlockAttached and OnTransactionConfirmed

Commit:
ui: Sync transactions when confirm/attached

Commit:
ui: Move on to the new architecture

Commit:
ui: Add hide button

Commit:
ui: Update transaction details page

- Add NavigationArrowBack icon widget to page common

- Remove button 'Hide' and add back button at page header

- Change transactionPage.details => transactionPage.txnInfo

Commit:
ui: Update construct method TransactionPage

Commit:
ui: Remove param transaction at TransactionPage method

oshorefueled commented on ui/handle_input.go:
```diff
@@ -660,3 +678,14 @@ func (win *Window) sortTransactions() {
+			}
+		}
+	}
+}
```
This be restructured to the new page implementation. The page should handle its own events.

oshorefueled commented on ui/transaction_page.go:
```diff
@@ -0,0 +1,338 @@
+	"fmt"
+	"os/exec"
+	"runtime"
+
```
The name of the page file should clearly depict what the page is about. This could be "transaction_details_page.go" and the list of transactions page could be "transactions_page.go"

oshorefueled commented on ui/transactions.go:
```diff
@@ -1,8 +1,10 @@
 package ui
 
 import (
+	"image"
```
Similar to the transactions details page, this should have a clear file name. It could be "transactions_page.go"

Commit:
ui: Update Transaction details

ui: Update Transaction details
- Change transaction.go -> transaction_details_page.go
- Layout the page with Navigation Tabs
- Add txsToTxsDetails at transactions.go
- Remove toTransactionDetails at widgets.go

Commit:
ui: Implement click on txs rows
- Add (*transactionsPage).updateTotransactionDetailsButtons()
- Add (*transactionsPage).goToTxnDetails()
- Add walletTransaction, toTxnDetails to transactionsPage struct

Commit:
ui: fix lint error

oshorefueled commented on ui/overview_page.go:
```diff
@@ -171,7 +171,6 @@ func (page overviewPage) recentTransactionsColumn() {
 					dcrlibwallet.ExtractDateOrTime(txn.Txn.Timestamp))),
 				status: theme.Body1(txn.Status),
 			}
-
```
you can just remove the overview_page.go entirely from the commit.

oshorefueled commented on wallet/txlistener.go:
```diff
@@ -18,24 +18,25 @@ type TxConfirmed struct {
 func (l *listener) OnBlockAttached(walletID int, blockHeight int32) {
 	l.Send <- SyncStatusUpdate{
 		Stage: BlockAttached,
+		ProcessTxn: NewBlock{
```
you can simply name this "BlockInfo" instead of "ProcessTxn"

oshorefueled commented on wallet/txlistener.go:
```diff
@@ -18,24 +18,25 @@ type TxConfirmed struct {
 	l.Send <- SyncStatusUpdate{
-		BlockConfirmed: TxConfirmed{
+		Stage: BlockConfirmed,
+		ProcessTxn: TxConfirmed{
```
You can name this "ConfirmedTxn" instead of "ProcessTxn"

Commit:
multi: Update listener
- Add BlockInfo, ConfirmedTxn -> listener.go
